File: sdks/python/src/memoturn/bedrock.py
"""Drop-in wrapper for a boto3 ``bedrock-runtime`` client — records ``converse``/
``converse_stream`` calls as generations. Only the standardized Converse API is
covered; ``invoke_model``/``invoke_model_with_response_stream`` are out of scope
since their request/response body shape varies per underlying model family."""
from __future__ import annotations

from typing import Any, Optional

from ._stream import _ABANDONED, _RecordingStream, _get
from .client import Memoturn, Trace
from .decorator import get_client

#: Model parameters copied onto the generation (allowlist — Bedrock's Converse API has
#: a small, stable inference-parameter set, unlike Gemini's larger/unstable config bag).
_ALLOWLIST = ("maxTokens", "temperature", "topP", "stopSequences")


def _map_usage(usage: Any) -> Optional[dict]:
    if not usage:
        return None
    out = {
        "promptTokens": _get(usage, "inputTokens"),
        "completionTokens": _get(usage, "outputTokens"),
        "totalTokens": _get(usage, "totalTokens"),
    }
    cache_read = _get(usage, "cacheReadInputTokens")
    if cache_read is not None:
        out["cacheReadTokens"] = cache_read
    cache_write = _get(usage, "cacheWriteInputTokens")
    if cache_write is not None:
        out["cacheCreationTokens"] = cache_write
    return out


def _build_input(kwargs: dict) -> tuple[Any, dict]:
    inference_config = kwargs.get("inferenceConfig") or {}
    model_parameters = {k: inference_config[k] for k in _ALLOWLIST if k in inference_config}
    system = kwargs.get("system")
    # Keep the system prompt alongside the messages when present (mirrors Anthropic's
    # own system+messages input shape).
    input_val = {"system": system, "messages": kwargs.get("messages")} if system else kwargs.get("messages")
    return input_val, model_parameters


def wrap_bedrock(client: Any, memoturn: Optional[Memoturn] = None, *, trace: Optional[Trace] = None) -> Any:
    """Patch ``client.converse``/``client.converse_stream`` to trace calls. Returns the
    same client. ``converse_stream`` is patched only if present.

    Streaming calls are recorded too: the ``stream`` key of the response dict is wrapped
    so events are forwarded unchanged to the caller while being accumulated into the same
    output/usage shape as a non-streaming call, and the generation is closed when the
    stream is exhausted, errors, or is abandoned (closed/garbage-collected/idle-timed-out
    before completion).
    """
    mt = memoturn or get_client()
    _patch_converse(client, mt, trace)
    _patch_converse_stream(client, mt, trace)
    return client


def _patch_converse(client: Any, mt: Memoturn, trace: Optional[Trace]) -> None:
    original = client.converse

    def converse(**kwargs: Any) -> Any:
        t = trace or mt.trace(name="bedrock.converse")
        input_val, model_parameters = _build_input(kwargs)
        gen = t.generation(
            name="bedrock.converse",
            model=kwargs.get("modelId"),
            provider="bedrock",
            input=input_val,
            modelParameters=model_parameters,
        )
        try:
            resp = original(**kwargs)
            gen.end(output=_get(_get(resp, "output"), "message"), usage=_map_usage(_get(resp, "usage")))
            return resp
        except Exception as e:  # noqa: BLE001
            gen.end(level="ERROR", statusMessage=str(e))
            raise

    client.converse = converse  # type: ignore[assignment]


class _BedrockAccumulator:
    """Accumulates ``converse_stream`` events (``contentBlockStart``/
    ``contentBlockDelta``/``metadata``) into the same ``(output, usage)`` shape the
    non-streaming path builds. Structurally mirrors ``_AnthropicAccumulator``: blocks are
    tracked by index, text deltas concatenate, and non-text deltas (e.g. ``toolUse``)
    merge generically."""

    def __init__(self) -> None:
        self._blocks: dict[int, dict] = {}
        self._usage: Any = None

    def add(self, event: dict) -> None:
        start = event.get("contentBlockStart")
        if start is not None:
            self._blocks[start["contentBlockIndex"]] = dict(start.get("start") or {})
        delta_evt = event.get("contentBlockDelta")
        if delta_evt is not None:
            i = delta_evt["contentBlockIndex"]
            block = self._blocks.setdefault(i, {})
            delta = delta_evt.get("delta") or {}
            if "text" in delta:
                block["text"] = block.get("text", "") + delta["text"]
            else:
                block.update(delta)
        metadata = event.get("metadata")
        if metadata is not None and metadata.get("usage") is not None:
            self._usage = metadata["usage"]

    def finalize(self) -> tuple[list, Optional[dict]]:
        return [self._blocks[i] for i in sorted(self._blocks)], _map_usage(self._usage)


def _patch_converse_stream(client: Any, mt: Memoturn, trace: Optional[Trace]) -> None:
    original = getattr(client, "converse_stream", None)
    if original is None:
        return

    def converse_stream(**kwargs: Any) -> Any:
        t = trace or mt.trace(name="bedrock.converseStream")
        input_val, model_parameters = _build_input(kwargs)
        gen = t.generation(
            name="bedrock.converseStream",
            model=kwargs.get("modelId"),
            provider="bedrock",
            input=input_val,
            modelParameters=model_parameters,
        )
        try:
            resp = original(**kwargs)
        except Exception as e:  # noqa: BLE001
            gen.end(level="ERROR", statusMessage=str(e))
            raise

        acc = _BedrockAccumulator()

        def on_chunk(event: dict) -> None:
            acc.add(event)

        def on_done(err: Any) -> None:
            output, usage = acc.finalize()
            if err is _ABANDONED:
                gen.end(level="WARNING", statusMessage="stream ended before completion", output=output, usage=usage)
            elif err is not None:
                gen.end(level="ERROR", statusMessage=str(err), output=output, usage=usage)
            else:
                gen.end(output=output, usage=usage)

        resp["stream"] = _RecordingStream(resp["stream"], on_chunk, on_done)
        return resp

    client.converse_stream = converse_stream  # type: ignore[assignment]

File: sdks/js/src/bedrock.ts
import type { Memoturn, MemoturnTrace } from "./client.js";
import { tapStream } from "./stream.js";

/** Model parameters worth recording — Bedrock's Converse API has a small, stable
 * `inferenceConfig` shape (unlike Gemini's large/evolving `config` bag), so this uses
 * the same allowlist philosophy as the Anthropic wrapper. */
const MODEL_PARAMETER_ALLOWLIST = ["maxTokens", "temperature", "topP", "stopSequences"] as const;

type WrapOptions = { trace?: MemoturnTrace; traceName?: string; streamTimeoutMs?: number };

/** Map Bedrock Converse usage (incl. prompt-cache tokens) to the recorded usage shape —
 * shared by the non-streaming path and the stream accumulator's `.finalize()`. */
function mapUsage(usage: any): Record<string, number> | undefined {
  if (!usage) return undefined;
  return {
    promptTokens: usage.inputTokens,
    completionTokens: usage.outputTokens,
    totalTokens: usage.totalTokens,
    ...(usage.cacheReadInputTokens != null ? { cacheReadTokens: usage.cacheReadInputTokens } : {}),
    ...(usage.cacheWriteInputTokens != null ? { cacheCreationTokens: usage.cacheWriteInputTokens } : {}),
  };
}

/** Build the recorded `input` (system + messages, mirroring Anthropic's own system+messages
 * shape, or bare messages when there's no system prompt) and the allowlisted `modelParameters`
 * extracted from `inferenceConfig`. */
function buildInput(input: any): { input: unknown; modelParameters: Record<string, unknown> } {
  const modelParameters: Record<string, unknown> = {};
  for (const key of MODEL_PARAMETER_ALLOWLIST) {
    const value = input?.inferenceConfig?.[key];
    if (value !== undefined) modelParameters[key] = value;
  }
  return {
    input: input?.system != null ? { system: input.system, messages: input?.messages } : input?.messages,
    modelParameters,
  };
}

/** Merge Bedrock ConverseStream events (`contentBlockStart`/`contentBlockDelta`/`metadata`)
 * into the same `{ output, usage }` shape the non-streaming path records — structurally
 * almost identical to `createAnthropicStreamAccumulator` in ./anthropic.ts. */
function createBedrockStreamAccumulator() {
  const blocks: Record<number, any> = {};
  let usage: any;

  return {
    add(event: any) {
      if (event?.contentBlockStart) {
        const { contentBlockIndex, start } = event.contentBlockStart;
        blocks[contentBlockIndex] = { ...start };
      }
      if (event?.contentBlockDelta) {
        const { contentBlockIndex, delta } = event.contentBlockDelta;
        const block = (blocks[contentBlockIndex] ??= {});
        if (delta?.text != null) {
          block.text = (block.text ?? "") + delta.text;
        } else if (delta) {
          // Non-text delta (e.g. toolUse, reasoningContent) — merge whatever fields it
          // carries as a best effort, same fallback the Anthropic accumulator uses.
          Object.assign(block, delta);
        }
      }
      if (event?.metadata?.usage) usage = event.metadata.usage;
    },
    finalize() {
      const output = Object.keys(blocks)
        .map(Number)
        .sort((a, b) => a - b)
        .map((i) => blocks[i])
        .filter(Boolean);
      return { output, usage: mapUsage(usage) };
    },
  };
}

/**
 * Drop-in wrapper for an AWS SDK v3 `BedrockRuntimeClient` — records `Converse`/
 * `ConverseStream` calls as memoturn generations (model, allowlisted `inferenceConfig`
 * params, usage incl. cache tokens, latency, output). **Only the standardized Converse
 * API is covered** — `InvokeModel`/`InvokeModelWithResponseStream` (raw, per-model-family
 * request/response bodies) are out of scope and pass through completely untouched; see
 * the README for details.
 *
 * Unlike every other wrapper in this package, AWS SDK v3 routes every operation through a
 * single `client.send(command)` call — there's no `client.converse(...)` method to
 * intercept. This wrapper proxies `.send` and inspects `command?.constructor?.name` (a
 * plain string comparison, not `instanceof` — this file never imports
 * `@aws-sdk/client-bedrock-runtime`) to decide whether to instrument the call.
 *
 *   const bedrock = wrapBedrock(new BedrockRuntimeClient({}), memoturn);
 *   await bedrock.send(new ConverseCommand({ modelId, messages }));
 *
 * Pass `{ trace }` to nest generations under an existing trace; otherwise each call gets
 * its own trace. Pass `{ streamTimeoutMs }` to override the idle-stream abandonment
 * backstop (default 120s).
 */
export function wrapBedrock<T extends object>(client: T, memoturn: Memoturn, options: WrapOptions = {}): T {
  return new Proxy(client, {
    get(target, prop, receiver) {
      const original = Reflect.get(target, prop, receiver);
      if (prop !== "send" || typeof original !== "function") return original;

      return async function send(command: any, ...rest: any[]) {
        const name = command?.constructor?.name;
        if (name !== "ConverseCommand" && name !== "ConverseStreamCommand") {
          return (original as (...args: any[]) => any).call(target, command, ...rest);
        }

        const input = command.input ?? {};
        const trace = options.trace ?? memoturn.trace({ name: options.traceName ?? "bedrock.converse" });
        const { input: recordedInput, modelParameters } = buildInput(input);
        const generation = trace.generation({
          name: name === "ConverseStreamCommand" ? "bedrock.converseStream" : "bedrock.converse",
          model: input.modelId,
          provider: "bedrock",
          modelParameters,
          input: recordedInput,
        });

        try {
          const response = await (original as (...args: any[]) => Promise<any>).call(target, command, ...rest);

          if (name === "ConverseCommand") {
            generation.end({ output: response?.output?.message, usage: mapUsage(response?.usage) });
            return response;
          }

          const accumulator = createBedrockStreamAccumulator();
          return {
            ...response,
            stream: tapStream(
              response.stream,
              {
                onChunk: (event: any) => accumulator.add(event),
                onDone: (err, reason) => {
                  if (reason === "error") {
                    generation.end({ level: "ERROR", statusMessage: String(err), ...accumulator.finalize() });
                  } else if (reason === "abandoned") {
                    generation.end({
                      level: "WARNING",
                      statusMessage: "stream ended before completion",
                      ...accumulator.finalize(),
                    });
                  } else {
                    generation.end({ ...accumulator.finalize() });
                  }
                },
              },
              { idleTimeoutMs: options.streamTimeoutMs },
            ),
          };
        } catch (err) {
          generation.end({ level: "ERROR", statusMessage: String(err) });
          throw err;
        }
      };
    },
  });
}
